bundles/org.openhab.binding.somfymylink/src/main/java/org/openhab/binding/somfymylink/internal/handler/SomfyMyLinkBridgeHandler.java

```diff
@@ -0,0 +1,478 @@
+        } catch (ExecutionException e) {
+            logger.debug("Exception while getting scene list. Message: {}", e.getMessage());
+            throw new SomfyMyLinkException("Problem getting scene list.", e);
+        }
+    }
+
+    public void commandShadeUp(String targetId) throws SomfyMyLinkException {
+        SomfyMyLinkConfiguration config = this.config;
+        if (config == null || StringUtils.isEmpty(config.systemId)) {
+            throw new SomfyMyLinkException("Config not setup correctly. System Id is not set.");
+        }
+
+        try {
+            SomfyMyLinkCommandShadeUp cmd = new SomfyMyLinkCommandShadeUp(targetId, config.systemId);
+            sendCommand(cmd);
+        } catch (SomfyMyLinkException e) {
+            logger.info("Error commanding shade up: {}", e.getMessage());
+            throw new SomfyMyLinkException("Error commanding shade up", e);
+        }
+    }
+
+    public void commandShadeDown(String targetId) throws SomfyMyLinkException {
+        SomfyMyLinkConfiguration config = this.config;
+        if (config == null || StringUtils.isEmpty(config.systemId)) {
+            throw new SomfyMyLinkException("Config not setup correctly. System Id is not set.");
+        }
+
+        try {
+            SomfyMyLinkCommandShadeDown cmd = new SomfyMyLinkCommandShadeDown(targetId, config.systemId);
+            sendCommand(cmd);
+        } catch (SomfyMyLinkException e) {
+            logger.info("Error commanding shade down: {}", e.getMessage());
+            throw new SomfyMyLinkException("Error commanding shade down", e);
+        }
+    }
+
+    public void commandShadeStop(String targetId) throws SomfyMyLinkException {
+        SomfyMyLinkConfiguration config = this.config;
+        if (config == null || StringUtils.isEmpty(config.systemId)) {
+            throw new SomfyMyLinkException("Config not setup correctly. System Id is not set.");
+        }
+
+        try {
+            SomfyMyLinkCommandShadeStop cmd = new SomfyMyLinkCommandShadeStop(targetId, config.systemId);
+            sendCommand(cmd);
+        } catch (SomfyMyLinkException e) {
+            logger.info("Error commanding shade stop: {}", e.getMessage());
+            throw new SomfyMyLinkException("Error commanding shade stop", e);
+        }
+    }
+
+    public void commandScene(Integer sceneId) throws SomfyMyLinkException {
+        SomfyMyLinkConfiguration config = this.config;
+        if (config == null || StringUtils.isEmpty(config.systemId)) {
+            throw new SomfyMyLinkException("Config not setup correctly. System Id is not set.");
+        }
+
+        try {
+            SomfyMyLinkCommandSceneSet cmd = new SomfyMyLinkCommandSceneSet(sceneId, config.systemId);
+            sendCommand(cmd);
+        } catch (SomfyMyLinkException e) {
+            logger.info("Error commanding scene run: {}", e.getMessage());
+            throw new SomfyMyLinkException("Error commanding scene run", e);
+        }
```

Comment:
It makes no sense to wrap a SomfyMyLinkException in another SomfyMyLinkException. Either don't catch it at all or rethrow the cought SomfyMyLinkException like so:

```suggestion
        try {
            SomfyMyLinkCommandSceneSet cmd = new SomfyMyLinkCommandSceneSet(sceneId, config.systemId);
            sendCommand(cmd);
        } catch (SomfyMyLinkException e) {
            logger.info("Error commanding scene run: {}", e.getMessage());
            throw e;
        }
```

Although the issue with this approach is that you are logging the error then rethrowing it, at which point it can be assumed that the caller will log it again result in multiple errors being logged even if it was all the same error.

When you catch an exception, it should be with the intent to handle it, and when you throw an exception it is done when you can't handle something. Logging is an appropriate way to handle exceptions so when you both log and throw another exception it indicates that you both handled and didn't handle the exception.

So I think the better solution here is to not catch the exception at all.
```suggestion
        sendCommand(new SomfyMyLinkCommandSceneSet(sceneId, config.systemId));
```

But there is a final problem with this approach that has gone unnoticed until now: `sendCommand` doesn't even throw a SomfyMyLinkException!
Since sendCommand is now asynchronous you need to do your exception handling asynchronously as well, and since the caller of these command methods can't handle exceptions asynchronously you can handle them here.
```suggestion
        sendCommand(new SomfyMyLinkCommandSceneSet(sceneId, config.systemId))
        .exceptionally(e -> {
            logger.info("Error commanding scene run: {}", e.getMessage())
            return null;
        });
```
Now while you could use the above suggestion I think my second suggestion is better as long as you make sure that exception logging is handled within the `sendCommand` method. Handling the exception in the `sendCommand` method would eliminate a lot of this redundant error logging logic by doing it one place.

Reply:
got it. ok i tidied this up quite a bit and opted for option 2.